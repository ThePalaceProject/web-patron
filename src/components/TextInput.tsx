// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck
import * as React from "react";

/**
 * A base component that can be easily customized within
 * the CPW design system
 */

const styles = {
  borderRadius: 1,
  border: "solid",
  px: 2,
  py: 1,
  fontSize: "0",
  variant: "text.body.regular",
  width: "100%",
  "&::placeholder": {
    fontStyle: "italic",
    color: "ui.gray.dark",
    fontSize: "-1"
  }
};

const TextInput = React.forwardRef<
  HTMLInputElement,
  React.InputHTMLAttributes<HTMLInputElement>
>(({ type = "text", ...props }, ref) => {
  return <input ref={ref} type={type} sx={styles} {...props} />;
});
TextInput.displayName = "TextInput";

export const TextArea = React.forwardRef<
  HTMLTextAreaElement,
  React.InputHTMLAttributes<HTMLTextAreaElement>
>((props, ref) => {
  return <textarea ref={ref} sx={styles} {...props} />;
});
TextArea.displayName = "TextArea";

export default TextInput;
