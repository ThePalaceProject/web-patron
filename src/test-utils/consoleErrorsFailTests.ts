import { jest } from "@jest/globals";
const consoleErrorSpy = jest.spyOn(global.console, "error");

// use this to make console errors fail tests. Useful when running in to issues.
export default function consoleErrorsFailTests() {
  consoleErrorSpy.mockImplementation((msg, ...opts) => {
    console.warn(msg, ...opts);
    throw new Error(msg);
  });
}
