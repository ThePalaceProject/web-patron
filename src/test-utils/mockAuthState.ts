import { AuthCredentials, OPDS1 } from "interfaces";
import Cookie from "js-cookie";
import { generateCredentials } from "utils/auth";
import { basicTokenAuthenticationUrl } from "./fixtures";

const mockCookie = Cookie as any;

export const creds: AuthCredentials = {
  token: "some-token",
  methodType: OPDS1.BasicAuthType
};

export const persistentUserCredentials = generateCredentials("1234", "pinpin");
export const firstToken = "IaMaBeArErToKeN";
export const tokenCreds1: AuthCredentials = {
  token: {
    basicToken: persistentUserCredentials,
    bearerToken: firstToken
  },
  authenticationUrl: basicTokenAuthenticationUrl,
  methodType: OPDS1.BasicTokenAuthType
};

export const newToken = "IaMaBeArErToKeN2";
export const tokenCreds2: AuthCredentials = {
  token: {
    basicToken: persistentUserCredentials,
    bearerToken: newToken
  },
  authenticationUrl: basicTokenAuthenticationUrl,
  methodType: OPDS1.BasicTokenAuthType
};

const str = JSON.stringify;

export default function mockAuthenticatedOnce(
  credentials: AuthCredentials | null = creds
) {
  mockCookie.get.mockReturnValueOnce(str(credentials));
}

export function mockAuthenticated(credentials: AuthCredentials | null = creds) {
  mockCookie.get.mockReturnValue(str(credentials));
}
