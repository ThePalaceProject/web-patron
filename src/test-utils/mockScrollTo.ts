import { jest } from "@jest/globals";
/**
 * JSdom doesn't include a stub for the scrollTo function, which we use inside of
 * Lane. Therefore we will stub it globally here
 */

Element.prototype.scrollTo = jest.fn();
window.scrollTo = jest.fn();

export default {};
