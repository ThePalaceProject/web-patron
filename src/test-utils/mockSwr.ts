import { jest } from "@jest/globals";
import { responseInterface } from "swr";

export function makeSwrResponse<T>(value?: Partial<responseInterface<T, any>>) {
  return {
    data: undefined,
    error: undefined,
    revalidate: jest.fn(),
    isValidating: false,
    mutate: jest.fn(),
    ...value
  };
}

export type MockSwr<T> = (value?: Partial<responseInterface<T, any>>) => void;
