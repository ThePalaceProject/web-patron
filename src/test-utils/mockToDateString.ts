import { jest } from "@jest/globals";
/**
 * This file mocks Date.prototype.toDateString so that it just returns a
 * stub string. This makes code using that function testable across timezones
 * since we can't reliably set the timezone of a node process to something specific.
 */
export const MOCK_DATE_STRING = "mock-date-string";
export const toDateStringSpy = jest
  .spyOn(Date.prototype, "toDateString")
  .mockImplementation(() => MOCK_DATE_STRING);
