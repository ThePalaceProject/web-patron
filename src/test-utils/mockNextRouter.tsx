// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck
import { jest } from "@jest/globals";
import * as React from "react";
import Router from "next/router";
import { RouterContext } from "next/dist/shared/lib/router-context.shared-runtime";
import { NextRouter } from "next/router";
import { libraryData } from "test-utils/fixtures";

/**
 * Mock for the next/Router import.
 */

export const mockPush = jest.fn().mockImplementation(async () => true);
Router.push = mockPush;

export const MockNextRouterContextProvider: React.FC<{
  router?: Partial<NextRouter>;
}> = ({ router = {}, children }) => {
  const {
    isReady = true,
    basePath = "",
    route = "",
    pathname = "",
    query = {
      // add the default library slug to the default query params
      library: libraryData.slug
    },
    // default as path is the home page
    asPath = `/${libraryData.slug}`,
    push = mockPush,
    replace = jest.fn().mockImplementation(async () => true),
    reload = jest.fn().mockImplementation(() => null),
    back = jest.fn().mockImplementation(() => null),
    prefetch = jest.fn().mockImplementation(async () => undefined),
    beforePopState = jest.fn().mockImplementation(() => null),
    isFallback = false,
    isLocaleDomain = false,
    isPreview = false,
    events = {
      on: () => null,
      off: () => null,
      emit: () => null
    }
  } = router;
  return (
    <RouterContext.Provider
      value={{
        isReady,
        basePath,
        route,
        pathname,
        query,
        asPath,
        push,
        replace,
        reload,
        back,
        prefetch,
        beforePopState,
        isFallback,
        isLocaleDomain,
        isPreview,
        events
      }}
    >
      {children}
    </RouterContext.Provider>
  );
};

export default Router;
