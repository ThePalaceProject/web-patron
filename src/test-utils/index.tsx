import * as React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import { beforeEach, expect, jest } from "@jest/globals";
import Adapter from "enzyme-adapter-react-16";
import { configure } from "enzyme";
import { LibraryData } from "../interfaces";
import "./mockScrollTo";
import * as fixtures from "./fixtures";
import serializer from "jest-emotion";
import { Provider as ReakitProvider } from "reakit";
import { MockNextRouterContextProvider } from "./mockNextRouter";
import { NextRouter } from "next/router";
import { enableFetchMocks } from "jest-fetch-mock";
import { LibraryProvider } from "components/context/LibraryContext";
import { BreadcrumbProvider } from "components/context/BreadcrumbContext";
import { UserContext, UserState } from "components/context/UserContext";
import { ThemeProvider } from "theme-ui";
import makeTheme from "theme";
import mockConfig from "test-utils/mockConfig";
import track from "analytics/track";
import "react-intersection-observer/test-utils";
import "test-utils/mockToDateString";

enableFetchMocks();
expect.addSnapshotSerializer(serializer);

// standard config mock
beforeEach(() => {
  mockConfig();
});

// mock the error tracker to prevent unnecessary console logs
export const mockTrackError = jest.fn();
track.error = mockTrackError;

export { fixtures };

// configure the enzyme adapter
configure({ adapter: new Adapter() });

/**
 * mock out the window.URL.createObjectURL since it isn't
 * available on jsdom
 */
const mockCreateObjectURL = jest.fn();
(global as any)["URL"].createObjectURL = mockCreateObjectURL;

type CustomRenderOptions = Parameters<typeof render>[1] & {
  router?: Partial<NextRouter>;
  library?: Partial<LibraryData>;
  user?: Partial<UserState>;
};
/**
 * Our custom render function wraps components in mocked versions of our
 * providers
 */
const customRender = (ui: any, options?: CustomRenderOptions) => {
  const library: LibraryData = {
    ...fixtures.libraryData,
    ...options?.library
  };
  const theme = makeTheme(library.colors);

  const user: UserState = {
    ...fixtures.user,
    ...options?.user
  };

  const AllTheProviders: React.FC = ({ children }) => {
    return (
      <MockNextRouterContextProvider router={options?.router}>
        <ThemeProvider theme={theme}>
          <ReakitProvider>
            <LibraryProvider library={library}>
              <UserContext.Provider value={user}>
                <BreadcrumbProvider>{children}</BreadcrumbProvider>
              </UserContext.Provider>
            </LibraryProvider>
          </ReakitProvider>
        </ThemeProvider>
      </MockNextRouterContextProvider>
    );
  };
  return {
    ...render(ui, { wrapper: AllTheProviders, ...options })
  };
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
