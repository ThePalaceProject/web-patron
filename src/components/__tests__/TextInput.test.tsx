/** @jsxRuntime classic */
/** @jsx jsx */
import { expect, test } from "@jest/globals";
import { jsx } from "theme-ui";
import * as React from "react";
import { render } from "../../test-utils";
import TextInput, { TextArea } from "../TextInput";

test("TextInput renders properly and passes styles through", () => {
  const utils = render(<TextInput sx={{ color: "red" }} />);
  expect(utils.container.firstChild).toMatchSnapshot();

  expect(utils.container.firstChild).toHaveStyle("color: red");
});

test("type textarea renders properly and passes styles throug", () => {
  const utils = render(<TextArea sx={{ color: "maroon" }} />);
  expect(utils.container.firstChild).toMatchSnapshot();
  expect(utils.container.firstChild).toHaveStyle("color: maroon");
});
