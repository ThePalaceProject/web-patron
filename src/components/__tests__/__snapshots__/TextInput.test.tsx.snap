// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput renders properly and passes styles through 1`] = `
.emotion-0 {
  border-radius: 2px;
  border: 1px solid #bdbdbd;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 4px;
  padding-bottom: 4px;
  font-size: 1rem;
  font-family: 'Open Sans',Helvetica,Arial,sans-serif;
  font-weight: 300;
  line-height: 1.5;
  width: 100%;
  color: red;
}

.emotion-0::-webkit-input-placeholder {
  font-style: italic;
  color: var(--theme-ui-colors-ui-gray-dark);
  font-size: 0.875rem;
}

.emotion-0::-moz-placeholder {
  font-style: italic;
  color: var(--theme-ui-colors-ui-gray-dark);
  font-size: 0.875rem;
}

.emotion-0:-ms-input-placeholder {
  font-style: italic;
  color: var(--theme-ui-colors-ui-gray-dark);
  font-size: 0.875rem;
}

.emotion-0::placeholder {
  font-style: italic;
  color: var(--theme-ui-colors-ui-gray-dark);
  font-size: 0.875rem;
}

<input
  class="emotion-0"
  type="text"
/>
`;

exports[`type textarea renders properly and passes styles throug 1`] = `
.emotion-0 {
  border-radius: 2px;
  border: 1px solid #bdbdbd;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 4px;
  padding-bottom: 4px;
  font-size: 1rem;
  font-family: 'Open Sans',Helvetica,Arial,sans-serif;
  font-weight: 300;
  line-height: 1.5;
  width: 100%;
  color: maroon;
}

.emotion-0::-webkit-input-placeholder {
  font-style: italic;
  color: var(--theme-ui-colors-ui-gray-dark);
  font-size: 0.875rem;
}

.emotion-0::-moz-placeholder {
  font-style: italic;
  color: var(--theme-ui-colors-ui-gray-dark);
  font-size: 0.875rem;
}

.emotion-0:-ms-input-placeholder {
  font-style: italic;
  color: var(--theme-ui-colors-ui-gray-dark);
  font-size: 0.875rem;
}

.emotion-0::placeholder {
  font-style: italic;
  color: var(--theme-ui-colors-ui-gray-dark);
  font-size: 0.875rem;
}

<textarea
  class="emotion-0"
/>
`;
